tests/python_package_test/test_arrow.py

```diff
@@ -432,3 +432,15 @@ def test_predict_ranking():
         num_boost_round=5,
     )
     assert_equal_predict_arrow_pandas(booster, data)
+
+
+def test_arrow_categorical():
+    data = generate_random_arrow_table(10, 10000, 42)
+    dataset = lgb.Dataset(
+        data,
+        label=generate_random_arrow_array(10000, 43, generate_nulls=False, values=np.arange(4)),
+        params=dummy_dataset_params(),
+        categorical_feature=["col_0"]
+    )
+    booster = lgb.train({"num_leaves": 7}, dataset, num_boost_round=5)
+    assert_equal_predict_arrow_pandas(booster, data)
```

Comment:
This PR is about ensuring that the feature names are derived correctly, but the test is just checking that predicting on `pandas` and `pyarrow` is the same.

Could you please modify the test such that the input data has custom feature names (not `col_0`, `col_1`, etc. similar to what LightGBM automatically assigns internally)? Like `red`, `black`, `green` or something (any random nonsense is fine).

And then explicitly test that `booster.feature_name()` returns the exact expected values?

That'd be a stronger test that this is working as expected.

Reply:
> This PR is about ensuring that the feature names are derived correctly

IIRC, I think this PR is really only about ensuring that `categorical_feature` is picked up correctly. Feature names were already stored correctly (and tested for) beforehand 🤔

Reply:
I don't think so. The issue says:

> the docs [say that] if feature_name="auto", a lightgbm.Dataset can infer feature names from the column names of a pyarrow table. It appears that this does not happen.

And this PR is title "Infer feature names from `pyarrow.Table`".

I also don't see any tests on the correct handling for feature names in https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_arrow.py (but please correct me if I'm wrong).

Reply:
> I also don't see any tests on the correct handling for feature names

I think this is implicitly covered by 
https://github.com/microsoft/LightGBM/blob/e0c34e7b2f68793bbf46854b85fd750be134446d/tests/python_package_test/test_arrow.py#L138-L141

Reply:
hmmm maybe, but that's related to what I'm asking for in this thread... more explicit tests.

Bugs like "feature names for both `pandas` and `pyarrow` tables are ignored" wouldn't be caught by a test like that. In other words, these two things are related but not identical:

* "handling of `pyarrow` tables is consistent with its handling of `pandas` dataframes"
* "handling of `pyarrow` tables is correct"

I'd prefer we explicitly test both.

Reply:
Yup, fair enough 👍🏼

Reply:
The `feature_name` were correctly assigned before. It's just that they were assigned *after* the check for `categorical_feature`'s, so this failed.

I had just copied tests from `test_pandas` and aligned them with other `test_arrow` tests. I can make the test check `feature_name()`